test/exchanges/oasis/Eth2DaiDirect.spec.js

```diff
@@ -0,0 +1,223 @@
+    expect(Object.keys(normalParams[2])).toEqual(['_bn']);
+    expect(normalParams[3]).toEqual(wethAddress);
+    expect(Object.keys(normalParams[4])).toEqual(['_bn']);
+
+    expect(ethParams.length).toEqual(4);
+    expect(ethParams[0]).toEqual(otcAddress);
+    expect(ethParams[1]).toEqual(wethAddress);
+    expect(ethParams[2]).toEqual(daiAddress);
+    expect(Object.keys(ethParams[3])).toEqual(['_bn']);
+
+    expect(buyAndPayEthParams.length).toEqual(4);
+    expect(buyAndPayEthParams[0]).toEqual(otcAddress);
+    expect(buyAndPayEthParams[1]).toEqual(daiAddress);
+    expect(Object.keys(buyAndPayEthParams[2])).toEqual(['_bn']);
+    expect(buyAndPayEthParams[3]).toEqual(wethAddress);
+
+    expect(createParams.length).toEqual(4);
+    expect(createParams[0]).toEqual(registryAddress);
+    expect(createParams[1]).toEqual(otcAddress);
+    expect(createParams[2]).toEqual(daiAddress);
+    expect(Object.keys(createParams[3])).toEqual(['_bn']);
+  });
+
+  test('set transaction options', () => {
+    const normalOptions = service._buildOptions(1, 'DAI', 'method');
+    const ethOptions = service._buildOptions(1, 'ETH', 'method');
+    const createOptions = service._buildOptions(1, 'ETH', 'create');
+    expect(Object.keys(normalOptions)).toEqual(['otc', 'dsProxy']);
+    expect(Object.keys(ethOptions)).toEqual(['otc', 'dsProxy', 'value']);
+    expect(Object.keys(createOptions)).toEqual(['otc', 'value']);
+  });
+});
+
+describe('values from otc', () => {
+  beforeAll(async () => {
+    await createDaiAndPlaceLimitOrder(service);
+  });
+
+  test('get buy amount', async () => {
+    const buyAmount = await service.getBuyAmount('WETH', 'DAI', '0.01');
+    expect(Object.keys(buyAmount)).toEqual(['_bn']);
+    expect(buyAmount.toString()).toEqual('500000000000000');
+  });
+
+  test('get minBuyAmount', async () => {
+    const limit = await service._minBuyAmount('WETH', 'DAI', '0.01');
+    expect(limit.toString()).toEqual('490000000000000');
+  });
+
+  test('get pay amount', async () => {
+    const payAmount = await service.getPayAmount('DAI', 'WETH', '0.01');
+    expect(Object.keys(payAmount)).toEqual(['_bn']);
+    expect(payAmount.toString()).toEqual('200000000000000000');
+  });
+
+  test('get maxPayAmount', async () => {
+    const limit = await service._maxPayAmount('DAI', 'WETH', '0.01');
+    expect(limit.toString()).toEqual('204000000000000000');
+  });
+});
+
+describe('trade with existing dsproxy', () => {
+  beforeEach(async () => {
+    if (!proxyAccount) {
+      proxyAccount = TestAccountProvider.nextAccount();
+    }
+    await setProxyAccount(service, proxyAccount);
+    if (!(await proxy())) await service.get('proxy').build();
+  });
+
+  describe('sell dai', () => {
+    beforeEach(async () => {
+      await createDaiAndPlaceLimitOrder(service);
+    });
+
+    test('sell all amount', async () => {
+      const order = await service.sell('DAI', 'WETH', '0.01');
+      expect(Object.keys(order)).toEqual(orderKeys);
```

Comment:
what could you do in these tests to more explicitly test that the actions have had the desired effects?

Reply:
`fillAmount` wasn't working reliably, and `fees` would change sometimes with the state of the network. I could test the balances before and after the order, but I was hesitant to add more time to the test. If I succeed in speeding it up, then I can test for those values.

Reply:
I fixed the issue with the logs from the otc contract and now test for the specific fill amount of each order.